File: apps/api/internal/models/rank.go
package models

import (
	"time"
)

type Rank int

const (
	RankScribe     Rank = 1
	RankChronicler Rank = 2
	RankCurator    Rank = 3
	RankPreserver  Rank = 4
	RankWarden     Rank = 5
	RankOracle     Rank = 6
)

func (r Rank) String() string {
	switch r {
	case RankScribe:
		return "Scribe"
	case RankChronicler:
		return "Chronicler"
	case RankCurator:
		return "Curator"
	case RankPreserver:
		return "Preserver"
	case RankWarden:
		return "Warden"
	case RankOracle:
		return "Oracle"
	default:
		return "Unknown"
	}
}

type UserRank struct {
	ID          int64     `db:"id" json:"id"`
	UserID      int64     `db:"user_id" json:"userId"`
	SeasonID    *int64    `db:"season_id" json:"seasonId,omitempty"` 
	CurrentRank Rank      `db:"current_rank" json:"currentRank"`
	CurrentElo  int       `db:"current_elo" json:"currentElo"`
	PeakRank    Rank      `db:"peak_rank" json:"peakRank"`
	CreatedAt   time.Time `db:"created_at" json:"createdAt"`
	UpdatedAt   time.Time `db:"updated_at" json:"updatedAt"`
}

File: apps/web/src/types/rank.ts
export enum Rank {
    Scribe = 1,
    Chronicler = 2,
    Curator = 3,
    Preserver = 4,
    Warden = 5,
    Oracle = 6
}

export interface UserRank {
    id: number;
    userId: number;
    seasonId?: number;
    currentRank: Rank;
    currentElo: number;
    peakRank: Rank;
    createdAt: string;
    updatedAt: string;
}

export const RankNames: Record<Rank, string> = {
    [Rank.Scribe]: "Scribe",
    [Rank.Chronicler]: "Chronicler",
    [Rank.Curator]: "Curator",
    [Rank.Preserver]: "Preserver",
    [Rank.Warden]: "Warden",
    [Rank.Oracle]: "Oracle"
};

export const RankThresholds: Record<Rank, number> = {
    [Rank.Scribe]: 0,
    [Rank.Chronicler]: 150,
    [Rank.Curator]: 400,
    [Rank.Preserver]: 800,
    [Rank.Warden]: 1500,
    [Rank.Oracle]: 2500
};
